core/contextmenu.js

```diff
@@ -301,6 +301,116 @@ Blockly.ContextMenu.blockCommentOption = function(block) {
   return commentOption;
 };
 
+/**
+ * Make a context menu option for undoing the most recent action on the
+ * workspace.
+ * @param {!Blockly.WorkspaceSvg} ws The workspace where the right-click
+ *     originated.
+ * @return {!Object} A menu option, containing text, enabled, and a callback.
+ * @package
+ */
+Blockly.ContextMenu.wsUndoOption = function(ws) {
+  return {
+    text: Blockly.Msg['UNDO'],
+    enabled: ws.hasUndoStack(),
+    callback: ws.undo.bind(ws, false)
+  };
+};
+
+/**
+ * Make a context menu option for redoing the most recent action on the
+ * workspace.
+ * @param {!Blockly.WorkspaceSvg} ws The workspace where the right-click
+ *     originated.
+ * @return {!Object} A menu option, containing text, enabled, and a callback.
+ * @package
+ */
+Blockly.ContextMenu.wsRedoOption = function(ws) {
+  return {
+    text: Blockly.Msg['REDO'],
+    enabled: ws.hasRedoStack(),
```

Comment:
Looking at these 3 options, seems like "enabled" is the the common thread, but you're getting at that in different ways (i.e. sometimes by passing the info in, other times by adding methods to workspace).  I'm wondering if it would be better to just pass the boolean in. That's almost what you're doing below with numTopBlocks. 
It also means you could just have a more generic addOption method that takes the text, enabled, and callback.

After reading a bit more code, it might work for collapse and expand too?

Reply:
That feels like it's pushing the code back into workspace, which is what I was trying to get rid of.  That's sort of the structure that the code has over there, just with inline creation of the objects.